src/record_ts.cpp

```diff
@@ -0,0 +1,1446 @@
+                    // v->literal == identity
+                    size = op.size;
+                    jitc_log(LogLevel::Debug,
+                             "jitc_memcpy_async(dst=%p, src= literal 0x%lx, "
+                             "size=%zu)",
+                             dst_rv.data, op.data, (size_t) size * tsize);
+                }
+
+                if (size != op.size)
+                    return false;
+
+                uint32_t replication_per_worker =
+                    size == 1u ? (64u / tsize) : 1u;
+                size_t new_size =
+                    size * (size_t) replication_per_worker * (size_t) workers;
+
+                dst_rv.alloc(backend, new_size, dst_info.vtype);
+
+                if (!dry_run)
+                    ts->memset_async(dst_rv.data, new_size, tsize, &op.data);
+
+                if (!dry_run && memcpy)
+                    ts->memcpy_async(dst_rv.data, src_ptr,
+                                     (size_t) size * tsize);
+
+                dst_rv.data_size = size * type_size[(uint32_t) dst_info.vtype];
+                // dst_rv.size = size;
+            };
+            break;
+            case OpType::Compress: {
+                ProfilerPhase profiler(pr_compress);
+
+                uint32_t dependency_index = op.dependency_range.first;
+
+                ParamInfo in_info  = this->dependencies[dependency_index];
+                ParamInfo out_info = this->dependencies[dependency_index + 1];
+
+                ReplayVariable &in_rv  = replay_variables[in_info.slot];
+                ReplayVariable &out_rv = replay_variables[out_info.slot];
+
+                uint32_t size = in_rv.size(in_info.vtype);
+                out_rv.alloc(backend, size, out_info.vtype);
+
+                if (dry_run)
+                    jitc_fail(
+                        "replay(): dry_run compress operation not supported!");
+
+                uint32_t out_size = ts->compress((uint8_t *) in_rv.data, size,
+                                                 (uint32_t *) out_rv.data);
+
+                out_rv.data_size =
+                    out_size * type_size[(uint32_t) out_info.vtype];
+            };
+            break;
+            case OpType::MemcpyAsync: {
+                ProfilerPhase profiler(pr_memset_async);
+
+                uint32_t dependency_index = op.dependency_range.first;
+                ParamInfo src_info = this->dependencies[dependency_index];
+                ParamInfo dst_info = this->dependencies[dependency_index + 1];
+
+                ReplayVariable &src_var = replay_variables[src_info.slot];
+                ReplayVariable &dst_var = replay_variables[dst_info.slot];
+
+                // size_t size = src_var.size(src_info.vtype);
+                jitc_log(LogLevel::Debug,
+                         "replay(): MemcpyAsync s%u <- s%u [%zu]",
+                         dst_info.slot, src_info.slot, src_var.data_size);
+
+                dst_var.alloc(backend, src_var.data_size);
```

Comment:
Could it be a problem that the memory is allocated even during a dry-run? (for this and other ops)
Will it be quickly freed again? We don't want to accidentally double memory usage because of dry runs.

Reply:
Ah right, I'll add a comment explaining it. `rv.alloc()` does not perform allocation when in dry run mode. It just sets the size (`data_size`) of the variable in that case.